In [7]:
import os
os.environ['TF_CPP_MIN_LOG_LEVEL'] = '3' 

import tensorflow as tf
from tensorflow import keras

import pandas as pd
import numpy as np

In [8]:
train_df = pd.read_csv('./data/train.csv')
np.random.shuffle(train_df.values)

print(train_df.head())

model = keras.Sequential([
	keras.layers.Dense(16, input_shape=(2,), activation='relu'),
    # keras.layers.Dropout(0.2),									# % of neurons (of previous layer) that will be dropped
	keras.layers.Dense(16, activation='relu'),						# adding layer, because quadratic example is more complex
	keras.layers.Dense(2, activation='sigmoid')])

model.compile(optimizer='adam', 
	          loss=keras.losses.SparseCategoricalCrossentropy(from_logits=True),
	          metrics=['accuracy'])

train_x = np.column_stack((train_df.x.values, train_df.y.values))

results = model.fit(train_x, train_df.color.values, batch_size=4, epochs=5)

# losses = results.history['loss']
# accuracies = results.history['accuracy']

          x          y  color
0 -1.268940   3.486986    0.0
1  4.178078  19.133821    0.0
2  0.505496   2.048246    0.0
3  2.563425   7.764043    0.0
4 -4.487521  21.198215    0.0
Epoch 1/5


/home/hubert/.local/lib/python3.10/site-packages/keras/src/backend.py:5727: UserWarning: "`sparse_categorical_crossentropy` received `from_logits=True`, but the `output` argument was produced by a Softmax activation and thus does not represent logits. Was this intended?
  output, from_logits = _get_logits(


1000/1000 [==============================] - 4s 4ms/step - loss: 0.5838 - accuracy: 0.7035
Epoch 2/5
1000/1000 [==============================] - 3s 3ms/step - loss: 0.3416 - accuracy: 0.8698
Epoch 3/5
1000/1000 [==============================] - 3s 3ms/step - loss: 0.2335 - accuracy: 0.9120
Epoch 4/5
1000/1000 [==============================] - 3s 3ms/step - loss: 0.1582 - accuracy: 0.9445
Epoch 5/5
1000/1000 [==============================] - 3s 3ms/step - loss: 0.1013 - accuracy: 0.9797


In [9]:
test_df = pd.read_csv('./data/test.csv')
test_x = np.column_stack((test_df.x.values, test_df.y.values))

print("EVALUATION")
evaluation = model.evaluate(test_x, test_df.color.values)

EVALUATION
32/32 [==============================] - 0s 3ms/step - loss: 0.0698 - accuracy: 0.9930


/home/hubert/.local/lib/python3.10/site-packages/keras/src/backend.py:5727: UserWarning: "`sparse_categorical_crossentropy` received `from_logits=True`, but the `output` argument was produced by a Softmax activation and thus does not represent logits. Was this intended?
  output, from_logits = _get_logits(


+ Increasing number of neurons from 4 to 16 increases accuracy from 65% to over 90%

+ Dropout did not improve the accuracy

+ Additional layers improved accuracy to 98%